### ISSUE TITLE
Kubernetes Client creation with Access tokens - Auto refresh tokens?

### ISSUE THREAD

**Issue by ajaykmis** [fabric8io/kubernetes-client#2963]
Hi, 

I am using fabric Kubernetes client to talk to AWS- EKS cluster. this is how I create the client right now: 
```
return new DefaultKubernetesClient(
        configBuilder
            .withMasterUrl(cluster.getEndpoint())
            .withClientCertData(cluster.getCertificateAuthority().getData())
            .withOauthToken(initialToken)
            .build()); 
```

The access token is generated using AWS "sts" service and is only valid for 60 secs. Is there a way to implement a custom authenticator so, that my client is always valid?
Right now every time I need K8Client, I have to generate a new access token. I can implement a caching mechanism to save the token and return the same client reference, but was wondering if there's a way to pass a custom method to the Kubernetes client which refreshes the token.

Any pointers will be helpful.

**Comment by manusa** [comment]
Could you share a redacted version of you .kube/config file.
For OIDC we implemented a token refresh mechanism. In this case (AWS-EKS), we don't know the URL for the token refresh endpoint, neither how to infer it.

**Comment by manusa** [comment]
Related to #3097, may be fixed by #3105

**Comment by digitalsanctum** [comment]
I have a similar struggle. I created an implementation of `OAuthTokenProvider` with the following:
```
override fun getToken(): String {
    val accountId = kubernetesClusterProperties.computeProvider.accountId!!
    val clusterName = kubernetesClusterProperties.clusterName
    val region = Region.of(kubernetesClusterProperties.region)

    log.info("Getting token for account '$accountId', cluster '$clusterName' in $region region")
    return try {
      val requestToSign = SdkHttpFullRequest
          .builder()
          .method(SdkHttpMethod.GET)
          .uri(getStsRegionalEndpointUri(region))
          .appendHeader("x-k8s-aws-id", kubernetesClusterProperties.clusterName)
          .appendRawQueryParameter("Action", "GetCallerIdentity")
          .appendRawQueryParameter("Version", "2011-06-15")
          .build()

      log.info("Request to sign: ${requestToSign}")

      val expirationDate: ZonedDateTime = ZonedDateTime.now().plusSeconds(60)

      val creds = getAwsCredentialsProviderForSession(accountId).resolveCredentials()
      log.info("creds: $creds")

      val presignerParams = Aws4PresignerParams.builder()
          .awsCredentials(creds)
          .signingRegion(region)
          .signingName("sts")
          .signingClockOverride(Clock.systemUTC())
          .expirationTime(expirationDate.toInstant())
          .build()

      log.info("Signing parameters: $presignerParams")

      val signedRequest = Aws4Signer.create().presign(requestToSign, presignerParams)
      log.info("Signed Request: $signedRequest")

      val signedRequestBytes = signedRequest.uri.toString().toByteArray()
      val encodedUrl: String = Base64.getUrlEncoder()
          .withoutPadding()
          .encodeToString(signedRequestBytes)
      val token = "k8s-aws-v1.$encodedUrl"
      log.info("Returning token: $token")
      token
    } catch (e: Exception) {
      val errorMessage = "Error generating an EKS authentication token for cluster: $clusterName"
      log.error(errorMessage, e)
      throw RuntimeException(errorMessage, e)
    }
  }
  ```
  and I'm configuring the `KubernetesClient` with something like this:
  ```
  val builder: ConfigBuilder = ConfigBuilder()
              .withMasterUrl(endpoint)
              .withCaCertData(caData)
          if (trustCerts != null) {
            builder.withTrustCerts(trustCerts)
          }
          if (oAuthTokenProvider != null) {
            builder.withOauthTokenProvider(oAuthTokenProvider)
            builder.withOauthToken(oAuthTokenProvider.token)
          }

          builder.build()
```
In my case, I'm getting the following error when attempting to call a cluster endpoint:
```
io.fabric8.kubernetes.client.KubernetesClientException: Operation: [create] for kind: [Deployment] with name: [widget] in namespace: [ns1-test] failed.
at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:64)
at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:72)
at io.fabric8.kubernetes.client.dsl.base.BaseOperation.create(BaseOperation.java:373)
at io.fabric8.kubernetes.client.utils.CreateOrReplaceHelper.createOrReplace(CreateOrReplaceHelper.java:53)
at io.fabric8.kubernetes.client.dsl.base.BaseOperation.createOrReplace(BaseOperation.java:411)
at io.fabric8.kubernetes.client.dsl.base.BaseOperation.createOrReplace(BaseOperation.java:86)
at io.fabric8.kubernetes.client.dsl.base.BaseOperation.createOrReplace(BaseOperation.java:394)
at io.fabric8.kubernetes.client.dsl.base.BaseOperation.createOrReplace(BaseOperation.java:86)
at com.nike.sa.pcs.sandbox.service.KubernetesServiceImpl.createDeployment(KubernetesServiceImpl.kt:293)
at com.nike.sa.pcs.sandbox.service.KubernetesServiceImpl.createDeployment$default(KubernetesServiceImpl.kt:235)
at com.nike.sa.pcs.sandbox.service.KubernetesServiceImpl.deployWidget(KubernetesServiceImpl.kt:149)
at com.nike.sa.pcs.sandbox.service.KubernetesServiceImpl.handleDeploymentEvent(KubernetesServiceImpl.kt:68)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
at org.springframework.context.event.ApplicationListenerMethodAdapter.doInvoke(ApplicationListenerMethodAdapter.java:344)
at org.springframework.context.event.ApplicationListenerMethodAdapter.processEvent(ApplicationListenerMethodAdapter.java:229)
at org.springframework.context.event.ApplicationListenerMethodAdapter.onApplicationEvent(ApplicationListenerMethodAdapter.java:166)
at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:176)
at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:169)
at org.springframework.context.event.SimpleApplicationEventMulticaster.lambda$multicastEvent$0(SimpleApplicationEventMulticaster.java:140)
at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.FileNotFoundException: /root/.kube/config (No such file or directory)
at java.io.FileInputStream.open0(Native Method)
at java.io.FileInputStream.open(FileInputStream.java:195)
at java.io.FileInputStream.<init>(FileInputStream.java:138)
at com.fasterxml.jackson.dataformat.yaml.YAMLFactory.createParser(YAMLFactory.java:343)
at com.fasterxml.jackson.dataformat.yaml.YAMLFactory.createParser(YAMLFactory.java:14)
at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3333)
at io.fabric8.kubernetes.client.internal.KubeConfigUtils.parseConfig(KubeConfigUtils.java:42)
at io.fabric8.kubernetes.client.utils.TokenRefreshInterceptor.intercept(TokenRefreshInterceptor.java:44)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
at io.fabric8.kubernetes.client.utils.ImpersonatorInterceptor.intercept(ImpersonatorInterceptor.java:68)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
at io.fabric8.kubernetes.client.utils.HttpClientUtils.lambda$createApplicableInterceptors$6(HttpClientUtils.java:284)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:229)
at okhttp3.RealCall.execute(RealCall.java:81)
at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:485)
at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:448)
at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleCreate(OperationSupport.java:263)
at io.fabric8.kubernetes.client.dsl.base.BaseOperation.handleCreate(BaseOperation.java:870)
at io.fabric8.kubernetes.client.dsl.base.BaseOperation.create(BaseOperation.java:365)
```
